query-engine/core/src/query_document/parser.rs

```diff
@@ -572,25 +723,205 @@ impl QueryDocumentParser {
+                        .collect(),
+                )
+            })
+            .collect::<Vec<_>>()
+    }
+
+    pub(crate) fn input_type_to_argument_description(
+        arg_name: String,
+        input_type: &InputType,
+    ) -> validation::ArgumentDescription {
+        validation::ArgumentDescription::new(arg_name, vec![to_simplified_input_type_name(input_type)])
+    }
+
+    pub(crate) fn argument_value_to_type_name(value: &ArgumentValue) -> String {
+        match value {
+            ArgumentValue::Scalar(pv) => prisma_value_to_type_name(pv),
+            ArgumentValue::Object(_) => "Object".to_string(),
+            ArgumentValue::List(v) => {
+                format!("({})", itertools::join(v.iter().map(argument_value_to_type_name), ", "))
+            }
+            ArgumentValue::FieldRef(_) => "FieldRef".to_string(),
+        }
+    }
+
+    fn prisma_value_to_type_name(pv: &PrismaValue) -> String {
+        match pv {
+            PrismaValue::String(_) => "String".to_string(),
+            PrismaValue::Boolean(_) => "Boolean".to_string(),
+            PrismaValue::Enum(_) => "Enum".to_string(),
+            PrismaValue::Int(_) => "Int".to_string(),
+            PrismaValue::Uuid(_) => "UUID".to_string(),
+            PrismaValue::List(v) => {
+                format!("({})", itertools::join(v.iter().map(prisma_value_to_type_name), ", "))
+            }
+            PrismaValue::Json(_) => "JSON".to_string(),
+            PrismaValue::Xml(_) => "XML".to_string(),
+            PrismaValue::Object(_) => "Object".to_string(),
+            PrismaValue::Null => "Null".to_string(),
+            PrismaValue::DateTime(_) => "DateTime".to_string(),
+            PrismaValue::Float(_) => "Float".to_string(),
+            PrismaValue::BigInt(_) => "BigInt".to_string(),
+            PrismaValue::Bytes(_) => "Bytes".to_string(),
+        }
+    }
+
+    pub(crate) fn to_simplified_input_type_name(typ: &InputType) -> String {
+        match typ {
+            InputType::Enum(_) => String::from("enum"),
+            InputType::List(o) => format!("{}[]", to_simplified_input_type_name(o.as_ref())),
+            InputType::Object(o) => o
+                .upgrade()
+                .map(|f| f.identifier.name())
+                .unwrap_or_else(|| String::from("Object")),
+            InputType::Scalar(s) => s.to_string(),
+        }
+    }
+
+    pub(crate) fn to_simplified_output_type_name(typ: &OutputType) -> String {
+        match typ {
+            OutputType::Enum(_) => String::from("enum"),
+            OutputType::List(o) => format!("{}[]", to_simplified_output_type_name(o)),
+            OutputType::Object(o) => o
+                .upgrade()
+                .map(|f| f.identifier.name())
+                .unwrap_or_else(|| String::from("Object")),
+            OutputType::Scalar(s) => s.to_string(),
+        }
+    }
+}
+#[derive(Debug, Clone, Default)]
+pub struct Path {
+    segments: Vec<String>,
+}
```

Comment:
Does this need to be public? (genuinely asking, I don't know)

Reply:
I think it can be `pub(crate)`

Reply:
[bb438ba](https://github.com/prisma/prisma-engines/pull/3784/commits/bb438baf41e8405acf01996eecff224d2bd7d462)